### PR TITLE
Remove flow typing from main branch

### PR DESCRIPTION
**Pull request by noti0na1**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Types.scala [hunk 1]
```diff
@@ -2353,6 +2397,34 @@ object Types {
...
 
+  /** A `TermRef` that, through flow-sensitive type inference, we know is non-null.
+   *  Accordingly, the `info` in its denotation won't be of the form `T|Null`.
+   *
+   *  This class is cached differently from regular `TermRef`s. Regular `TermRef`s use the
+   *  `uniqueNameTypes` map in the context, while these non-null `TermRef`s use
+   *  the generic `uniques` map. This is so that regular `TermRef`s can continue to use
+   *  a "fast path", since non-null `TermRef`s are not very common.
+   */
+  final class NonNullTermRef(prefix: Type, designator: Designator) extends TermRef(prefix, designator) {
```
Comment (abeln) [inline compiler/src/dotty/tools/dotc/core/Types.scala]: This should go in the flow typing PR.

File: docs/docs/internals/explicit-nulls.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+The "explicit nulls" feature (enabled via a flag) changes the Scala type hierarchy
+so that reference types (e.g. `String`) are non-nullable. We can still express nullability
+with union types: e.g. `val x: String|Null = null`.
+
+The implementation of the feature in dotty can be conceptually divided in several parts:
+  1. changes to the type hierarchy so that `Null` is only a subtype of `Any`
+  2. a "translation layer" for Java interop that exposes the nullability in Java APIs
+  3. a "magic" `JavaNull` type (an alias for `Null`) that is recognized by the compiler and
+     allows unsound member selections (trading soundness for usability)
+  4. a module for "flow typing", so we can work more naturally with nullable values
```
Comment (abeln) [inline docs/docs/internals/explicit-nulls.md]: Let's remove this.

File: docs/docs/internals/explicit-nulls.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+      1. if `this` is not a nullable union, it's returned unchanged.
+      2. if `this` is a union, then it's re-arranged so that all the `Null`s are to the right of all
+         the non-`Null`s.
+  - `stripNull` syntactically strips nullability from `this`: e.g. `String|Null => String`. Notice this
+     works only at the "top level": e.g. if we have an `Array[String|Null]|Null` and we call `stripNull`
+     we'll get `Array[String|Null]` (only the outermost nullable union was removed).
+  - `stripAllJavaNull` is like `stripNull` but removes _all_ nullable unions in the type (and only works
+     for `JavaNull`). This is needed when we want to "revert" the Java nullification function.
+
+Flow Typing
```
Comment (abeln) [inline docs/docs/internals/explicit-nulls.md]: Let's remove this section.

File: docs/docs/internals/explicit-nulls.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    For example, to do FlowTyping on if expressions:
+      * we type the condition
+      * we give the typed condition to the FlowTyper and obtain a pair of sets of paths `(ifTrue, ifFalse)`.
+        We type the `then` branch with the `ifTrue` facts, and the else branch with the `ifFalse` facts.
+      * profit
+
+Flow typing also introduces two new abstractions: `NonNullTermRef` and `ValDefInBlockCompleter`.
+
+#### NonNullTermRef
```
Comment (abeln) [inline docs/docs/internals/explicit-nulls.md]: Remove

File: docs/docs/internals/explicit-nulls.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+This is a new type of `TermRef` (path-dependent type) that, whenever its denotation is updated, makes sure
+that the underlying widened type is non-null. It's defined in `Types.scala`. A `NonNullTermRef` is identified by `computeDenot` whenever the denotation is updated, and then we call `stripNull` on the widened type.
+
+To use the flow-typing information, whenever we see a path that we know must be non-null (in `typedIdent` or
+`typedSelect`), we replace its `TermRef` by a `NonNullTermRef`.
+
+#### ValDefInBlockCompleter
```
Comment (abeln) [inline docs/docs/internals/explicit-nulls.md]: Remove

File: docs/docs/reference/other-new-features/explicit-nulls.md [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+  /** Pass a String, or null to signal a special case */
+  def foo(s: String): Unit = ???
+}
+
+// User code in explicit-null world
+val o: Old = ???
+o.foo(null.asInstanceOf[String]) // ok: cast will succeed at runtime
+```
+
+## Flow Typing
```
Comment (abeln) [inline docs/docs/reference/other-new-features/explicit-nulls.md]: Remove

**Review by abeln — CHANGES_REQUESTED** [review]
Thanks for doing this! I found a few instances where we still referenced flow typing material, so those need to be removed.

**Comment by noti0na1** [comment]
@abeln Removed

**Review by abeln — APPROVED** [review]

**Pull request merged**